Make base compile on linux

ReplayCoding opened the pull request:
### Related Issue
TODO

### Implementation
This PR fixes a number of issues with compiling TC2 on linux. It currently runs, and can load maps. It uses parts of #616, #486, #603.
#### Build Instructions
```sh
podman run -v $PWD:/src -it registry.gitlab.steamos.cloud/steamrt/sniper/sdk
# this enters a container shell
cd /src
./src/build.sh
# exit the container
./game_clean/link.sh
./game/start_tf2.sh
```
### Screenshots
<!-- Add screenshots if applicable -->

### Checklist
<!-- You MUST answer "yes" to all of these to open a pull request -->
<!-- To tick a checkbox, place an 'x' in it, like so: [x] -->
- [x] No other PRs implement this idea.
- [x] This PR does not introduce any regressions.
- [X] I certify that this PR is my own entirely original work, or I have permission from and have attributed the relevant authors.
- [x] I have agreed to and signed this project's [Contributor License Agreement](https://cla-assistant.io/mastercomfig/team-comtress-2).
- [x] This PR only contains changes to the engine and/or core game framework
- [x] This PR targets the `community` branch. (???)

<!-- You do NOT have to answer "yes" to the following, but please mark them if relevant -->
<!-- To tick a checkbox, place an 'x' in it, like so: [x] -->
- [ ] This change has been filed as an issue.
- [x] No other PRs address this.
- [x] This PR is as minimal as possible.
- [x] This PR has been built and locally tested on at least one platform.
- [ ] This PR has been tested on all platforms, on both a dedicated server and on a listen server.

### Testing Checklist
<!-- You do not have to test on all platforms to open a pull request -->
|         |            Client             |            Server             | Version                     |
|---------|:-----------------------------:|:-----------------------------:|-----------------------------|
| Windows | <!-- Built, Tested or N/A --> | <!-- Built, Tested or N/A --> | <!-- e.g. Windows 10 -->    |
|   Linux | Built, Tested | <!-- Built, Tested or N/A --> | 6.1.7-xanmod1 #1-NixOS SMP PREEMPT_DYNAMIC Tue Jan  1 00:00:00 UTC 1980 |
|  Mac OS | <!-- Built, Tested or N/A --> | <!-- Built, Tested or N/A --> | <!-- e.g. Catalina -->      |

This was built under the steam "scout" and "sniper" runtime.

### Caveats
- ~While the game does run, there is a bug that will cause a crash whenever a weapon is inspected.~ fixed!

### Alternatives
- Don't support linux builds at all
- Rewrite the build system to use something like cmake

melvyn2 committed:
Fix sysctl.h usage

melvyn2 committed:
Clean up build process output

revert gendbg changes

melvyn2 committed:
Add old steamapi dylibs to game_clean

hyperventilation commented:
i am joe biden and this lgtm

melvyn2 commented:
Mind adding the following to commits which add build.sh & other scripts?
```
Co-authored-by: melvyn2 <melvyn2@brcok.tk>
Co-authored-by: HurricanePootis <eatabakhsh04@protonmail.com>
Co-authored-by: interrupt <int-10h@protonmail.com>
```

ReplayCoding commented:
> Mind adding the following to commits which add build.sh & other scripts?

Done, I've also added this to a couple of other commits

mastercoms reviewed (changes requested):
What's with this freetype2 change?
Also there are a lot of weird Linux specific changes that don't make a lot of sense from my perspective. Please revise this patch to be less unique to TC2. Attempt to follow the current state of TF2 as closely as possible, rather than doing what may be optimal. This will reduce the possibility of regressions and allow us to test things better.

ReplayCoding commented:
> What's with this freetype2 change?
TF2 already uses the system freetype libraries, and the headers that were in common don't actually build, so I just elected to use the system headers.

ReplayCoding committed:
add scripts to build repo

build.sh contains references to external libs which aren't added in this
commit.

Co-authored-by: melvyn2 <melvyn2@brcok.tk>
Co-authored-by: HurricanePootis <eatabakhsh04@protonmail.com>
Co-authored-by: interrupt <int-10h@protonmail.com>

ReplayCoding committed:
use system freetype headers

TF2 already uses the system library, and these headers are broken
anyways.

ReplayCoding committed:
fix case-sensitivity issues

linux paths are case-sensitive, some code doesn't expect this

ReplayCoding committed:
fix compile error in utlblockmemory.h

Without this, gcc will complain with the following error:
/src/src/public/tier1/utlblockmemory.h:139:2: error: there are no arguments to 'swap' that depend on a template parameter, so a declaration of 'swap' must be available [-fpermissive]
  139 |  swap( m_nBlocks, mem.m_nBlocks );
      |  ^~~~

ReplayCoding committed:
make gendbg.sh & ccache executable

ReplayCoding committed:
fix STEAM_RUNTIME_PATH on non-chroot builds

Use the steam runtime containers instead, they're easier to setup

ReplayCoding committed:
add some stuff to gitignore

contains things that will be added in a later commit

ReplayCoding committed:
fix protobuf builds

removed some files, and added them to the gitignore. these files will
get regenerated automatically, and if they're in the repo, it will show
up on "git status".

ReplayCoding committed:
fix a compile error caused by using -> on a non-pointer

I'm not sure if this will compile on windows, so *maybe* we should add a
ifdef check for linux. But, this should be correct.
/src/src/public/gcsdk/msgprotobuf.h:383:8: error: base operand of '->' has non-pointer type 'CMsgProtoBufHeader'
  383 |   Hdr()->set_client_session_id( nSessionIDClient );
      |        ^~

ReplayCoding committed:
use MIN macro in place of std::min

this is technically a problem, but it's what valve probably intended
/src/src/vgui2/vgui_surfacelib/linuxfont.cpp:505:43: error: no matching function for call to 'min(int&, unsigned int&)'
  505 |   int Width = min( rgbaWide, bitmap.width );

ReplayCoding committed:
add libedit

src/thirdparty/libedit-3.1/configure.ac is modified from the original
files, to make it work on modern autotools.

ReplayCoding committed:
add precompiled libraries

these are taken from the leaked source code

ReplayCoding committed:
fix a vphysics crash due to it not being able to find certain functions

failed to dlopen vphysics.so error=/home/user/Projects/tf2_stuff/cloned/team-comtress-2/game/bin/vphysics.so: undefined symbol: __wrap_freopen

ReplayCoding committed:
fix vpc crashing at runtime

VPC will segfault unless it's built as 32bit

ReplayCoding committed:
add working tcmalloc

ReplayCoding committed:
Don't run gensym/strip on static libraries

without this, static libraries such as tier1 won't contain debugging
info, making it harder to debug.

ReplayCoding committed:
Fix crash caused by target id panels

I've honestly got no idea what the root cause is, but this fixes it

ReplayCoding commented:
Reposting from the #tf2-patches channel on discord:
> If anyone wants to write a "proper" (if that even exists) fix for the target id panel crashes, feel free. I'm tired of trying to figure out what it is, and https://github.com/ReplayCoding/tf2-patches/commit/e05bc50c596b402a22f8e15e790059e03d244688 works fine.

mastercoms commented on src/game/protobuf_include.vpc:
```diff
@@ -12,7 +12,7 @@ $Project
 		$Libexternal $SRCDIR\thirdparty\protobuf-2.6.1\bin\osx32\libc++\libprotobuf [$OSX32]
 		$Libexternal $SRCDIR\thirdparty\protobuf-2.6.1\bin\win32\2013\staticcrt\release\libprotobuf [$WIN32]
-		$Libexternal $SRCDIR\thirdparty\protobuf-2.6.1\bin\linux32\libprotobuf [$LINUX32]
+		$Libexternal $SRCDIR\thirdparty\protobuf-2.6.1\src\.libs\libprotobuf [$LINUX32]
```
What's this change?

ReplayCoding replied:
This is where autotools puts build outputs, the previous path is specific to valve builds. Doing this is just easier, and is how thirdparty deps are referenced in a couple of other places, so it's more consistent.

mastercoms commented on src/game/shared/schemainitutils.h:
```diff
@@ -25,14 +25,17 @@
 		else															\
 		{																\
-#ifndef VALVE_PURE														\
+/* fixme(replaycoding):
```
Could you fix this?

ReplayCoding replied:
Yes... Is what I would say if this actually changed anything, reverting this will still build on sniper. :woman_shrugging: (maybe it failed to build on soldier & scout?)

ReplayCoding replied:
According to https://stackoverflow.com/questions/10074520/why-compiler-complain-about-this-macro-declaration, this isn't actually possible to fix without moving VALVE_PURE to outside the macro (ew, we're not doing that), so i'm just going to leave it commented out, and remove the fixme. (i'll add a note explaining why it doesn't work)

mastercoms commented on src/devtools/makefile_base_posix.mak:
```diff
@@ -575,7 +577,7 @@ $(GAMEOUTPUTFILE): $(OUTPUTFILE)
 		$(P4_EDIT_START) $(IMPORTLIBRARY) $(P4_EDIT_END) && \
 		mkdir -p `dirname $(IMPORTLIBRARY)` > /dev/null && \
-		cp -v $(OUTPUTFILE) $(IMPORTLIBRARY); \
+		cp $(OUTPUTFILE) $(IMPORTLIBRARY); \
```
Why was -v removed?

ReplayCoding replied:
Makes build output quieter

ReplayCoding replied:
I've reverted this, since I didn't fully change every instance of `cp`, no reason to have these be special

mastercoms commented on src/devtools/makefile_base_posix.mak:
```diff
@@ -557,10 +557,12 @@ $(GAMEOUTPUTFILE): $(OUTPUTFILE)
 	$(QUIET_PREFIX) -$(P4_EDIT_START) $(GAMEOUTPUTFILE)$(SYM_EXT) $(P4_EDIT_END);
-	$(QUIET_PREFIX) $(GEN_SYM) $(GAMEOUTPUTFILE); 
-	$(QUIET_PREFIX) -$(STRIP) $(GAMEOUTPUTFILE);
+	$(QUIET_PREFIX) -if [ "$(CONFTYPE)" != "lib" ]; then\
```
Could you explain this part?

ReplayCoding replied:
Without this, debug info from static libs won't show up in the final binaries.

mastercoms commented on src/devtools/makefile_base_posix.mak:
```diff
@@ -541,7 +541,7 @@ endif
 $(LIB_File): $(OTHER_DEPENDENCIES) $(OBJS) 
 	$(QUIET_PREFIX) -$(P4_EDIT_START) $(LIB_File) $(P4_EDIT_END); 
-	$(QUIET_PREFIX) $(AR) $(LIB_File) $(OBJS) $(LIBFILES);
+	$(QUIET_PREFIX) $(AR) rcs $(LIB_File) $(OBJS) $(LIBFILES);
```
Explain this part? You defined AR as `ar rcs` above

mastercoms commented on src/devtools/makefile_base_posix.mak:
```diff
@@ -421,7 +421,7 @@ else
 
 ifneq "$(origin VALVE_NO_AUTO_P4)" "undefined"
-	P4_EDIT_START = chmod -R +w
+	P4_EDIT_START = chmod -fR +w
```
Why -f?

ReplayCoding replied:
I got rid of this (`chmod -f`) in exchange for another change. This is to silence chmod errors when a file doesn't exist, which the new commit fixes by moving the invocations to *after* when they're built. But since I don't know what P4_EDIT_START does when perforce stuff is enabled, I'm not sure if that is a good idea... It may just be best to use this (`chmod -f`) instead

ReplayCoding replied:
I've decided to go back to using `chmod -f`, since all that does is silence errors. Since the errors don't happen on rebuilds, only fresh builds, i'm going to assume it's just an artifact of how the perforce stuff works, and not anything to worry about.

mastercoms commented on src/devtools/makefile_base_posix.mak:
```diff
@@ -183,7 +183,7 @@ endif
 # If not specified by environment, use steam runtime compilers + in-tree ccache
 ifneq ($(filter default undefined,$(origin AR)),)
-	AR = $(STEAM_RUNTIME_PATH)/bin/ar crs
+	AR = $(STEAM_RUNTIME_PATH)/bin/ar rcs 
```
Why changed the order here?

mastercoms commented on src/devtools/gendbg.sh:
```diff
@@ -36,9 +36,9 @@ if [ "$OUTFILEDIR" != "$INFILEDIR" ]; then
 fi
 
-pushd "$INFILEDIR"	
+pushd "$INFILEDIR" > /dev/null
```
Why redir?

ReplayCoding replied:
push & popd print the directory stack whenever they're invoked, this just silences them

mastercoms replied:
Why do they need to be silenced?

ReplayCoding replied:
It's not really important, but without this, the build will print the following whenever a shared lib is copied:
```
/src/game/bin /src/src/engine
/src/src/engine
```
It's not particularly important, so I'll revert it if you want.

ReplayCoding replied:
Reverted

mastercoms commented on build.sh:
```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env bash
```
Should be in src

ReplayCoding replied:
Should run.sh & link.sh be there too, or just this?

mastercoms replied:
all, just like windows scripts

mastercoms commented on src/tier1/KeyValues.cpp:
```diff
@@ -2607,7 +2607,15 @@ bool KeyValues::WriteAsBinary( CUtlBuffer &buffer )
 		case TYPE_NONE:
 			{
-				dat->m_pSub->WriteAsBinary( buffer );
+				if (dat->m_pSub)
```
I'm not sure if I like this fix

ReplayCoding replied:
It's not great, but it *is* the correct fix. The previous (valve) version was depending on the function being able to handle a null `this`. (which is UB, and caused gcc to optimise it out)

mastercoms reviewed (commented):
Still some problematic changes, also why was empty ccache added?

ReplayCoding commented:
> also why was empty ccache added?
Just github being funny, it's a perms change to make it executable.

ReplayCoding committed:
fix error when building static libs

ReplayCoding committed:
fix VGui buttons not being clickable on newer GCC versions

copied nearly verbatim from CSGO, thanks to ficool2 for pointing out
there was a better fix

mastercoms reviewed (changes requested):
Could you make the shell scripts consistent with how Windows operates? For example, why is there not a creategameprojects.sh? https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/creategameprojects

Why is run.sh not in game_clean? Why is it not in the start_server/start_tf2 format?

Ristovski commented:
Does this only build with some specific gcc version?

```
gcc --version
gcc (Gentoo 12.2.1_p20230121-r1 p10) 12.2.1 20230121
```

```
In file included from src/libc_override_glibc.h:64,
                 from src/libc_override.h:78,
                 from src/tcmalloc.cc:157,
                 from src/debugallocation.cc:79:
src/libc_override_gcc_and_weak.h:56:33: error: ISO C++17 does not allow dynamic exception specifications
   56 | void* operator new(size_t size) throw (std::bad_alloc)
      |                                 ^~~~~
src/libc_override_gcc_and_weak.h:60:35: error: ISO C++17 does not allow dynamic exception specifications
   60 | void* operator new[](size_t size) throw (std::bad_alloc)
      |                                   ^~~~~
cc1plus: note: unrecognized command-line option '-Wno-reserved-user-defined-literal' may have been intended to silence earlier diagnostics
make: *** [Makefile:4083: libtcmalloc_debug_la-debugallocation.lo] Error 1
make: *** Waiting for unfinished jobs....
```

ReplayCoding commented:
It's only been tested on the steam runtimes, which go up to GCC 10.3.0 on the sniper runtime.

Ristovski commented:
> It's only been tested on the steam runtimes, which go up to GCC 10.3.0 on the sniper runtime.

gcc 12 defaults to C++17, thus this will require setting `-std=gnu++14` explicitly (which is what gcc 10 defaults to)

mastercoms commented on game_clean/copy/start_tf2.sh:
```diff
@@ -0,0 +1,10 @@
+ARGS="-steam -game tf -insecure -novid -nojoy -nosteamcontroller -nohltv -particles 1 -noborder -particle_fallback 2 -dev -nobreakpad -console"
+
+export LAUNCH_PREFIX="$HOME/.steam/bin32/steam-runtime/run.sh"
+LAUNCH_PREFIX+=" $(command -v mangohud || true)"
```
Remove mangohud from here. Instead, allow users to customize their `LAUNCH_PREFIX` (script should add on `run.sh`).

mastercoms commented on game_clean/copy/start_tf2.sh:
```diff
@@ -0,0 +1,10 @@
+set -e
+cd "$(dirname "$0")"
+
+ARGS="-steam -game tf -insecure -novid -nojoy -nosteamcontroller -nohltv -particles 1 -noborder -particle_fallback 2 -dev -nobreakpad -console"
```
Remove `-nohltv -particles 1 -noborder -dev -console`

mastercoms commented on src/build.sh:
```diff
@@ -0,0 +1,108 @@
+
+# shellcheck disable=SC2086   # we want arguments to be split
+devtools/bin/vpc_linux /define:WORKSHOP_IMPORT_DISABLE /define:SIXENSE_DISABLE /define:NO_X360_XDK \
+				/define:RAD_TELEMETRY_DISABLED /define:DISABLE_ETW /retail /tf ${VPC_FLAGS} "+${VPC_GROUP}" /mksln games
```
Remove `/define:DISABLE_ETW /retail`, add `/nofpo`.

ReplayCoding replied:
`/nofpo` is not functional on linux, release builds always have frame pointer optimisations turned off.

mastercoms replied:
Yea, as is the case on Windows too :) but in case we patch it in the future, I don't want any surprises!!

mastercoms commented on src/game/shared/schemainitutils.h:
```diff
@@ -25,14 +25,17 @@
+/* note(replaycoding):
+ * this ifndef is invalid syntax, according to
+ * https://stackoverflow.com/questions/10074520/why-compiler-complain-about-this-macro-declaration */ \
+/* #ifndef VALVE_PURE														*/ \
```
This needs to be removed, or if you really can't figure it out, then remove the whole commented out part.

mastercoms commented on src/vpc_scripts/groups.vgc:
```diff
@@ -133,8 +133,8 @@ $Group "game"
 	"datamodel"
-	"dedicated"
-	"dedicated_main"
+//	"dedicated"
```
This is not needed on Windows, please figure out another solution for this or why dedicated can't be part of the same build.

ReplayCoding replied:
Reverted. dedicated builds fine, but it will fail to launch due to the game failing to load `SDLMgrInterface001` in `CMaterialSystem::Connect`. The code is inside of `#if !defined(DEDICATED)`, which won't get defined unless the DEDICATED macro is defined in VPC

mastercoms reviewed (changes requested):
It's looking pretty good, nice work and progress!

mastercoms commented:
Also, you should update create_clean.sh to also support linux

ReplayCoding committed:
Add linux install instructions to README

ReplayCoding committed:
Disable null-pointer check optimization in gcc

In certain situations, gcc will remove instances of code that check if a
pointer is null, if it thinks that the pointer can never be null.
However, many of these pointers *can* be null in TF2, so this
optimization can cause crashes.

ReplayCoding reviewed (commented):
*non-windows :trollface:
